### PR TITLE
Basic operator setup

### PR DESCRIPTION
**Pull request by alexander-demichev**
This PR does basic operator setup and mostly carries code structure from CCCMO.

1. Setup go modules
2. Add makefile
3. Add main.go
4. Setup cluster operator controller
5. Add cluster api feature gate

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+all: build
+
+verify: fmt vet lint
+
+# Run tests
+test: generate verify manifests unit
+
+unit:
+	hack/unit-tests.sh
```
Comment (lobziik) [inline Makefile:21]: Broken target

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+# Build operator binaries
+build: operator
+
+operator:
+	go build -o bin/cluster-controller-manager-operator cmd/cluster-cluster-api-operator/main.go
```
Comment (lobziik) [inline Makefile:27]: broken target

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+# Run against the configured Kubernetes cluster in ~/.kube/config
+run: verify manifests
+	go run cmd/cluster-cluster-api-operator/main.go
```
Comment (lobziik) [inline Makefile:31]: broken

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+# Run go mod
+.PHONY: vendor
+vendor:
+	go mod tidy
+	go mod vendor
+	go mod verify
+
+# Generate code
+generate: controller-gen
+	$(CONTROLLER_GEN) object:headerFile="hack/boilerplate.go.txt" paths="./..."
```
Comment (lobziik) [inline Makefile:61]: no hack dir in there. seems broken.

File: cmd/cluster-api-operator/main.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+var (
+	scheme   = runtime.NewScheme()
+	setupLog = ctrl.Log.WithName("setup")
+
+	leaderElectionConfig = config.LeaderElectionConfiguration{
+		LeaderElect:   true,
+		LeaseDuration: util.LeaseDuration,
+		RenewDeadline: util.RenewDeadline,
+		RetryPeriod:   util.RetryPeriod,
+		ResourceName:  "cluster-api-leader",
```
Comment (lobziik) [inline cmd/cluster-api-operator/main.go:31]: would prefer more specific name there.

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+# Build the docker image
+.PHONY: image
+image:
+	docker build -t ${IMG} .
+
+# Push the docker image
+.PHONY: push
+push:
+	docker push ${IMG}
```
Comment (lobziik) [inline Makefile:71]: Not everyone use docker. Can we wrap it?

Aslo, there is no dockerfile . Does this targets make sense?

File: Makefile (hunk 1 above)
Comment (alexander-demichev) [reply Makefile:21]: I know it's broken, my plan is to make a follow-up PR with missing tasks. I didn't want to add too many changes here.

File: Makefile (hunk 6 above)
Comment (alexander-demichev) [reply Makefile:71]: I will add a Dockerfile in a follow-up and yes I can default to podman as in other MAPI components.

File: cmd/cluster-api-operator/main.go (hunk 5 above)
Comment (alexander-demichev) [reply cmd/cluster-api-operator/main.go:31]: do you have any ideas?

File: Makefile (hunk 1 above)
Comment (lobziik) [reply Makefile:21]: This pr contains 1kk LOC btw...

File: cmd/cluster-api-operator/main.go (hunk 5 above)
Comment (lobziik) [reply cmd/cluster-api-operator/main.go:31]: `cluster-api-operator-leader` or `cluster-api-meta-operator-leader`

File: Makefile (hunk 1 above)
Comment (alexander-demichev) [reply Makefile:32]: It's all vendor directory :D

File: Makefile (hunk 6 above)
Comment (lobziik) [reply Makefile]: Then, i would suggest to remove targets which is not needed / does not work now and add it by necessity.

File: pkg/controllers/common_consts.go [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+const (
+	DefaultManagedNamespace = "openshift-cluster-api"
+
+	infrastructureResourceName = "cluster"
+	clusterOperatorName        = "cluster-api"
+	operatorVersionKey         = "operator"
+	externalFeatureGateName    = "cluster"
+
+	// ClusterAPIEnabled is the name of the external cloud provider feature gate.
```
Comment (lobziik) [inline pkg/controllers/common_consts.go:11]: doesn't looks actual

File: pkg/controllers/operator_status.go [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+func (r *ClusterOperatorReconciler) setStatusAvailable(ctx context.Context) error {
+	co, err := r.getOrCreateClusterOperator(ctx)
+	if err != nil {
+		klog.Errorf("Unable to set cluster operator status available: %v", err)
+		return err
+	}
+
+	conds := []configv1.ClusterOperatorStatusCondition{
+		newClusterOperatorStatusCondition(configv1.OperatorAvailable, configv1.ConditionTrue, ReasonAsExpected,
+			fmt.Sprintf("Cluster Cloud Controller Manager Operator is available at %s", r.ReleaseVersion)),
```
Comment (lobziik) [inline pkg/controllers/operator_status.go:37]: >Cluster Cloud Controller Manager Operator

**Comment by lobziik** [comment]
/lgtm
/hold
Want to leave others a room to review.

**Review by elmiko — COMMENTED** [review]
generally looks good to me
/lgtm

**Pull request merged**